lib/jxl/enc_cluster.cc

```diff
@@ -121,16 +147,20 @@ void FastClusterHistograms(const std::vector<Histogram>& in,
   for (size_t i = 0; i < in.size(); i++) {
     if ((*histogram_symbols)[i] != max_histograms) continue;
     size_t best = 0;
-    float best_dist = HistogramDistance(in[i], (*out)[best]);
-    for (size_t j = 1; j < out->size(); j++) {
+    float best_dist = std::numeric_limits<float>::max();
+    for (size_t j = 0; j < out->size(); j++) {
+      if (j < prev_histograms && HasNewSymbol(in[i], (*out)[j])) continue;
```

Comment:
This is an optimization, but since we do not AddHistogram below, the distance in the case of j < prev_histogram should just be given by the coding cost of the new histogram with the old one - which, interestingly, should be infinity if there is a new symbol, so HasNewSymbol should not be necessary.

Otherwise, histograms such as the flat histogram are somewhat too likely to be chosen, as they have typically very few symbols compared to the actual histogram, so HistogramDistance assumes they will move much closer to the actual histogram.

Reply:
Yes, good point. I want to evaluate separately so I added a TODO for now.